### PR TITLE
UI test exploration of messaging

### PR DESCRIPTION
**Pull request by jhugman**
This PR adds UI testing to the message notifications.

It tests if notifications are wired up to the message notification worker, and if that is wired up to the message system, and if that is wired up to Nimbus.

Additional tests for each message surface as it is added should be implemented. 

Separate tests are also implemented for the validity of the JSON as typed into the experimenter console.

cc @jrbenny35 

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation

Used by GitHub Actions.

### CODE REVIEW

**Comment by jrbenny35** [comment]
I would've expected these to have some sort of ```UiSelector``` calls in them, but I don't see any. How are you ensuring the messages show up on the UI?

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingTest.kt [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        )
+        activityTestRule.finishActivity()
+        FxNimbus.initialize { hardcodedNimbus }
+        activityTestRule.launchActivity(null)
+
+        mDevice.openNotification()
+        notificationShade {
+            val data = FxNimbus.features.messaging.value().messages["test-default-browser-notification"]
+            verifySystemNotificationExists(data!!.title!!)
+            verifySystemNotificationExists(data.text)
```
Comment (jhugman) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingTest.kt:179]: This detects when the notification is shown. 

It is working in Android API < O, but no later.

I don't know where the permission prompt is supposed to show in later versions. Does @rocketsroger or @rahulsainani know?

Comment (jrbenny35) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingTest.kt:179]: What about for the homescreen message?

Comment (rocketsroger) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingTest.kt:179]: Any device with Android 13 (Tieramisu) or above will need to user to opt-in (by showing the permission prompt first) before showing the notification.

Comment (jhugman) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingTest.kt:179]: Ok, fixed it!

It turned out to be that the `MessageNotificationWorker` wasn't ensuring that the marketing channel existed. 

@rocketsroger Should this be called earlier (e.g. in `showNotificationPermissionPromptIfRequired()` for `SDK < TIRAMISU`)?

File: fenix/app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt [hunk 2]
```diff
@@ -94,7 +102,7 @@ class MessageNotificationWorker(
 
         return createBaseNotification(
             context,
-            MARKETING_CHANNEL_ID,
+            ensureMarketingChannelExists(applicationContext),
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:105]: 👍

File: fenix/app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt [hunk 3]
```diff
@@ -148,21 +156,36 @@ class MessageNotificationWorker(
...
-                .build()
-
             val instanceWorkManager = WorkManager.getInstance(context)
-            instanceWorkManager.enqueueUniquePeriodicWork(
-                MESSAGE_WORK_NAME,
-                // We want to keep any existing scheduled work.
-                ExistingPeriodicWorkPolicy.KEEP,
-                messageWorkRequest,
-            )
+            if (pollingInterval > 0) {
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:160]: Drive by - don't we should be adding conditional checks just for tests - perhaps we could pass in the work request and work manager?

File: fenix/app/src/main/java/org/mozilla/fenix/nimbus/HardcodedNimbusFeatures.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * Placeholder class for a similar class in Application Services.
+ */
+class HardcodedNimbusFeatures(
+    override val context: Context,
+    private val features: Map<String, JSONObject>,
+) : FeaturesInterface {
+    private val exposureCounts = mutableMapOf<String, Int>()
+
+    constructor(context: Context, vararg pairs: Pair<String, JSONObject>) : this(
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/nimbus/HardcodedNimbusFeatures.kt:26]: pairs > features?

File: fenix/app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt [hunk 5]
```diff
@@ -49,6 +52,11 @@ class MessageNotificationWorker(
     override fun doWork(): Result {
         GlobalScope.launch(Dispatchers.IO) {
             val context = applicationContext
+            val nm = NotificationManagerCompat.from(applicationContext)
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:55]: could use the stored 'context'

**Comment by jhugman** [comment]
Closed in favour of https://github.com/mozilla-mobile/firefox-android/pull/1181.

**Pull request closed without merging**